index.go

```diff
@@ -0,0 +1,171 @@
+package opensearchtools
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+)
+
+// Index defines a method which knows how to make an OpenSearch [Index] request.
+// It should be implemented by a version-specific executor.
+//
+// [Index]: https://opensearch.org/docs/latest/api-reference/index-apis/index/
+type Index interface {
+	Index(ctx context.Context, req *IndexRequest) (OpenSearchResponse[IndexResponse], error)
+}
+
+type IndexRequest struct {
+	Operation IndexOperation
+	Routing   string
+}
+
+type IndexOp string
+
+const (
+	// IndexCreate creates an index and returns an error otherwise.
+	IndexCreate IndexOp = "create"
+
+	// IndexDelete delete an index if it exists and returns an error otherwise.
+	IndexDelete IndexOp = "delete"
+
+	// IndexGet gets an index if it exists and returns an error otherwise.
+	IndexGet IndexOp = "get"
+
+	// IndexExists checks the existence of an index
+	IndexExists IndexOp = "exists"
+)
+
+// IndexOperation various different index operations to be executed
+type IndexOperation struct {
+	Operation IndexOp
+	Indices   []string
+	Config    *indexRequestConfig
+	Doc       RoutableDoc
```

Comment:
I'm a little confused by this field. I don't think it should be a routable doc? Doesn't only create need a json body?

Reply:
I think raw json. So maybe just `io.Reader`. I'm not sure we want to make an entire index mapping entity at this point.

Reply:
I think I agree with @cs-marshall .  Was the intent to use `RoutableDoc` specifically because it already had the index field?  If so, then maybe I could be convinced that it makes sense here.

Reply:
No, the index is part of the request body itself. The index mapping doesn't match the same as other `Doc` entities.

Reply:
Yeah I'll update it with the raw json or simply io.Reader as Marshall suggested.